File: tests/vm/runtime_client.go
package main

import (
	"fmt"
	"io"
	"net"
	"os"
	"strings"
	"time"
)

func main() {
	if len(os.Args) != 2 {
		panic("usage: runtime-client tcp|udp4|udp6")
	}
	mode := os.Args[1]
	if mode == "tcp" {
		tcp()
	} else {
		udp(mode)
	}
	fmt.Printf("go-netgo-%s-ok\n", mode)
}

func tcp() {
	conn, err := net.DialTimeout("tcp", "fixture.test:18080", 5*time.Second)
	if err != nil {
		panic(err)
	}
	defer conn.Close()
	conn.SetDeadline(time.Now().Add(5 * time.Second))
	if _, err := io.WriteString(conn, "GET / HTTP/1.0\r\nHost: fixture.test\r\n\r\n"); err != nil {
		panic(err)
	}
	body, err := io.ReadAll(conn)
	if err != nil || !strings.HasSuffix(string(body), "fixture-v4") {
		panic(fmt.Sprintf("unexpected TCP response: %q (%v)", body, err))
	}
}

func udp(mode string) {
	network, target, expected := "udp4", "127.0.0.1:18082", "udp-v4:runtime-go"
	if mode == "udp6" {
		network, target, expected = "udp6", "[::1]:18083", "udp-v6:runtime-go"
	} else if mode != "udp4" {
		panic("unknown mode: " + mode)
	}
	peer, err := net.ResolveUDPAddr(network, target)
	if err != nil {
		panic(err)
	}
	conn, err := net.DialUDP(network, nil, peer)
	if err != nil {
		panic(err)
	}
	defer conn.Close()
	conn.SetDeadline(time.Now().Add(5 * time.Second))
	if _, err := conn.Write([]byte("runtime-go")); err != nil {
		panic(err)
	}
	response := make([]byte, 128)
	n, source, err := conn.ReadFromUDP(response)
	if err != nil || string(response[:n]) != expected || source.String() != peer.String() {
		panic(fmt.Sprintf("unexpected UDP response: %q from %v (%v)", response[:n], source, err))
	}
}

File: tests/vm/runtime_client.rs
use std::io::{Read, Write};
use std::net::{TcpStream, UdpSocket};
use std::time::Duration;

fn main() {
    let mode = std::env::args()
        .nth(1)
        .expect("usage: runtime-client tcp|udp4|udp6");
    match mode.as_str() {
        "tcp" => tcp(),
        "udp4" => udp(false),
        "udp6" => udp(true),
        _ => panic!("unknown mode: {mode}"),
    }
    println!("rust-{mode}-ok");
}

fn tcp() {
    let mut stream = TcpStream::connect(("fixture.test", 18080)).expect("connect TCP fixture");
    stream
        .set_read_timeout(Some(Duration::from_secs(5)))
        .expect("set TCP timeout");
    stream
        .write_all(b"GET / HTTP/1.0\r\nHost: fixture.test\r\n\r\n")
        .expect("write HTTP request");
    let mut response = Vec::new();
    stream.read_to_end(&mut response).expect("read HTTP response");
    assert!(response.ends_with(b"fixture-v4"), "unexpected TCP response");
}

fn udp(ipv6: bool) {
    let (bind, target, payload, expected) = if ipv6 {
        (
            "[::]:0",
            "[::1]:18083",
            b"runtime-rust".as_slice(),
            b"udp-v6:runtime-rust".as_slice(),
        )
    } else {
        (
            "0.0.0.0:0",
            "127.0.0.1:18082",
            b"runtime-rust".as_slice(),
            b"udp-v4:runtime-rust".as_slice(),
        )
    };
    let socket = UdpSocket::bind(bind).expect("bind UDP socket");
    socket.connect(target).expect("connect UDP fixture");
    socket
        .set_read_timeout(Some(Duration::from_secs(5)))
        .expect("set UDP timeout");
    socket.send(payload).expect("send UDP payload");
    let mut response = [0_u8; 128];
    let received = socket.recv(&mut response).expect("receive UDP response");
    assert_eq!(&response[..received], expected, "unexpected UDP response");
    assert_eq!(socket.peer_addr().expect("read UDP peer").to_string(), target);
}
